clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1804,21 +1816,25 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
     int Dims = static_cast<int>(
         AccTy->getTemplateArgs()[1].getAsIntegral().getExtValue());
     int Info = getAccessTarget(AccTy) | (Dims << 11);
-    Header.addParamDesc(SYCLIntegrationHeader::kind_accessor, Info, CurOffset);
+    Header.addParamDesc(SYCLIntegrationHeader::kind_accessor, Info, CurOffset,
+                        getNumOpenCLParams(AccTy));
     return true;
   }
-
+ 
   bool handleSyclSamplerType(FieldDecl *FD, QualType FieldTy) final {
     const auto *SamplerTy = FieldTy->getAsCXXRecordDecl();
     assert(SamplerTy && "Sampler type must be a C++ record type");
     CXXMethodDecl *InitMethod = getMethodByName(SamplerTy, InitMethodName);
     assert(InitMethod && "sampler must have __init method");
-
     // sampler __init method has only one argument
+    assert((InitMethod->param_size() == 1) &&
+           "sampler __init method should have only one argument");
     const ParmVarDecl *SamplerArg = InitMethod->getParamDecl(0);
     assert(SamplerArg && "sampler __init method must have sampler parameter");
-
-    addParam(FD, SamplerArg->getType(), SYCLIntegrationHeader::kind_sampler);
+    Header.addParamDesc(SYCLIntegrationHeader::kind_sampler,
```

Comment:
This change isn't necessary, since we know the size is 1.  If Sampler ever changes in size this change would have to be made though.  Also, by making this change, you lose some of the logic from addParam.